cli/src/commands/upload.command.ts

```diff
@@ -105,170 +115,333 @@ export class UploadOptionsDto {
   album? = false;
   albumName? = '';
   includeHidden? = false;
+  concurrency? = 4;
 }
 
 export class UploadCommand extends BaseCommand {
-  uploadLength!: number;
+  api!: ImmichApi;
 
   public async run(paths: string[], options: UploadOptionsDto): Promise<void> {
-    const api = await this.connect();
+    this.api = await this.connect();
 
-    const formatResponse = await api.getSupportedMediaTypes();
-    const crawlService = new CrawlService(formatResponse.image, formatResponse.video);
+    console.log('Crawling for assets...');
+    const files = await this.getFiles(paths, options);
 
-    const inputFiles: string[] = [];
-    for (const pathArgument of paths) {
-      const fileStat = await fs.promises.lstat(pathArgument);
-      if (fileStat.isFile()) {
-        inputFiles.push(pathArgument);
-      }
+    if (files.length === 0) {
+      console.log('No assets found, exiting');
+      return;
     }
 
-    const files: string[] = await crawlService.crawl({
-      pathsToCrawl: paths,
-      recursive: options.recursive,
-      exclusionPatterns: options.exclusionPatterns,
-      includeHidden: options.includeHidden,
-    });
+    const assetsToCheck = files.map((path) => new Asset(path));
 
-    files.push(...inputFiles);
+    const { newAssets, duplicateAssets, rejectedAssets } = await this.checkAssets(
+      assetsToCheck,
+      options.concurrency ?? 4,
+    );
 
-    if (files.length === 0) {
-      console.log('No assets found, exiting');
+    const totalSizeUploaded = await this.upload(newAssets, options);
+    const messageStart = options.dryRun ? 'Would have' : 'Successfully';
+    if (newAssets.length === 0) {
+      console.log('All assets were already uploaded, nothing to do.');
+    } else {
+      console.log(`${messageStart} uploaded ${newAssets.length} assets (${byteSize(totalSizeUploaded)})`);
+    }
+
+    if (!options.album && !options.albumName) {
+      return;
+    }
+
+    const { createdAlbumCount, updatedAssetCount } = await this.updateAlbums(
+      [...newAssets, ...duplicateAssets, ...rejectedAssets],
```

Comment:
Side note that I think the distinction between duplicate and rejected assets is a bit murky. I was expecting existing assets to all show up under `duplicateAssets` and for `rejectedAssets` to only be for files that were an unsupported format, etc. But existing assets also show up as `reject`.

Reply:
Oops, this was just a bug in how I was checking the status